barretenberg/cpp/src/barretenberg/commitment_schemes/shplonk/shplonk.hpp

```diff
@@ -206,67 +203,235 @@ template <typename Curve> class ShplonkVerifier_ {
                 G_commitment_constant += scaling_factor * opening_pair.evaluation;
-
                 // [G] -= ρʲ / ( r − xⱼ )⋅[fⱼ]
                 G_commitment -= commitment * scaling_factor;
-
                 current_nu *= nu;
             }
-
             // [G] += G₀⋅[1] = [G] + (∑ⱼ ρʲ ⋅ vⱼ / ( r − xⱼ ))⋅[1]
             G_commitment += g1_identity * G_commitment_constant;
         }
+        // Return opening pair (z, 0) and commitment [G]
+        return { { z_challenge, Fr(0) }, G_commitment };
+    };
+
+    /**
+    Shplonk verifier optimized to verify gemini opening claims.
+
+    This method receives commitments to all prover polynomials, their claimed evaluations, the sumcheck
+    challenge, a challenge \f$ \rho \f$ aimed to batch the commitments to prover polynomials, a challenge \f$ r \f$ for
+    the Gemini opening claims, and the Gemini claims. The latter is a tuple of a vector of powers of \f$ r \f$, a vector
+    of evaluations of Gemini fold polynomials at \f$ - r,
+    - r^2, \ldots, - r^{2^{d-1}} \f$ where \f$ d \f$ is the log_circuit_size, and a vector of commitments to the Gemini
+    fold polynomials.
+
+    The verifier receives the challenges required in Shplonk and gradually populates the vectors of commitments and
+    scalars that will be multiplied at the very end. In the recursive setting, a batch_mul of size (NUM_ALL_ENTITIES +
+    log_circuit_size + 2) is performed. In the native setting, these operations are performed sequentially.
+     *
+     */
+    static OpeningClaim<Curve> verify_gemini(
+        Commitment g1_identity,
+        RefSpan<Commitment> f_commitments,
+        RefSpan<Commitment> g_commitments,
+        auto claimed_evaluations,
+        std::vector<Fr>& multivariate_challenge,
+        const Fr& rho,
+        const Fr& gemini_r,
+        //  Fr a_0_pos,
+        std::tuple<std::vector<Fr>, std::vector<Fr>, std::vector<Commitment>> claims,
+        auto& transcript)
+    {
+        // (r, r^2, ... , r^{2^{d-1}}), where d = log circuit size
+        auto& r_squares = std::get<0>(claims);
+        // (A_0(-r), A_1(-r^2), ... , A_{d-1}(-r^{2^{d-1}}))
+        auto& gemini_evaluations = std::get<1>(claims);
+        // (com(A_1), com(A_2), ... , com(A_{d-1}))
+        auto& gemini_commitments = std::get<2>(claims);
+        const size_t num_claims = gemini_commitments.size();
+        // get Shplonk batching challenge
+        const Fr shplonk_batching_challenge = transcript->template get_challenge<Fr>("Shplonk:nu");
+        // quotient commitment for the batched opening claim
+        auto Q_commitment = transcript->template receive_from_prover<Commitment>("Shplonk:Q");
+        // get Shplonk opening point z, it used to check that the evaluation claims and the correctness of the batching
+        const Fr z_challenge = transcript->template get_challenge<Fr>("Shplonk:z");
+        // accumulator for scalar that will be multiplied by [1]_1
+        auto constant_term_accumulator = Fr(0);
```

Comment:
Please don't use this pattern; rather use `Fr constant_term_accumulator{ 0 };`. We should only instantiate objects with `auto` if we have a very compelling reason to do so. Types help us to understand what things are!